service-runner.js

```diff
@@ -143,6 +146,24 @@ ServiceRunner.prototype.updateConfig = function updateConfig(conf) {
     }
 };
 
+ServiceRunner.prototype._checkHeartbeat = function() {
+    var self = this;
+    self.interval = setInterval(function() {
+        if (!self._shuttingDown && !self._inRollingRestart) {
+            Object.keys(cluster.workers).forEach(function(workerId) {
+                var worker = cluster.workers[workerId];
+                var lastBeat = self.workerHeartbeatTime[worker.process.pid];
+                if (!lastBeat || (!lastBeat.killed && new Date() - lastBeat.time > self.config.timeout)) {
+                    self._logger.log('error/service-runner/master',
+                        'worker ' + worker.process.pid + ' stopped sending heartbeats, killing.');
+                    self._stopWorker(workerId);
+                    // Don't need to respawn a worker, it will be restarted upon 'exit' event
+                }
+            });
+        }
+    }, self.config.timeout);
```

Comment:
Hm, not really. We should still check if the last beat arrived more than `timeout` milliseconds ago. It's just that we are checking it more often.

The important thing here is that if/when the master and/or worker is super-busy, sending or checking the beat message might get scheduled way after the interval we have set. If that happens in the worker, we should obviously kill it, but if that happens in the master, it shouldn't start killing off workers because of it.